File: core/src/main/java/com/toocol/ssh/core/mosh/core/network/Transport.java
package com.toocol.ssh.core.mosh.core.network;

import com.toocol.ssh.core.mosh.core.proto.InstructionPB;
import com.toocol.ssh.core.mosh.core.statesnyc.CompleteTerminal;
import com.toocol.ssh.core.mosh.core.statesnyc.State;
import com.toocol.ssh.core.mosh.core.statesnyc.UserEvent;
import com.toocol.ssh.core.mosh.core.statesnyc.UserStream;
import com.toocol.ssh.core.term.core.Term;
import com.toocol.ssh.utilities.functional.DiffThread;
import com.toocol.ssh.utilities.execeptions.NetworkException;
import com.toocol.ssh.utilities.log.Loggable;
import com.toocol.ssh.utilities.utils.Timestamp;
import io.vertx.core.datagram.DatagramPacket;
import io.vertx.core.datagram.DatagramSocket;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedDeque;

@SuppressWarnings("all")
public final class Transport<RemoteState extends State> implements Loggable {
    public final Addr addr;
    private final RemoteState state;
    private final TransportFragment.Pool receivePool = new TransportFragment.Pool();
    private final TransportFragment.FragmentAssembly fragments = new TransportFragment.FragmentAssembly(receivePool);
    private final List<TimestampedState<RemoteState>> receiveStates = new ArrayList<>();
    private final Queue<InstructionPB.Instruction> instQueue = new ConcurrentLinkedDeque<>();
    private TransportSender<UserStream> sender;
    private Connection connection;

    public Transport(String serverHost, int port, String key, RemoteState initalRemoteState) {
        this.addr = new Addr(serverHost, port, key);
        this.state = initalRemoteState;
        receiveStates.add(new TimestampedState<>(Timestamp.timestamp(), 0, initalRemoteState));
    }

    public void connect(DatagramSocket socket) {
        this.connection = new Connection(this.addr, socket);
        this.sender = new TransportSender<>(new UserStream(), this.connection);
        this.sender.setSendDelay(1);
        this.receivePool.init();
        // tell the server the size of the terminal
        pushBackEvent(new UserEvent.Resize(Term.WIDTH, Term.HEIGHT));
    }

    @DiffThread
    public void receivePacket(DatagramPacket datagramPacket) {
        byte[] bytes = this.connection.recvOne(datagramPacket.data().getBytes());
        TransportFragment.Fragment fragment = receivePool.getObject().setData(bytes);
        if (fragments.addFragment(fragment)) {
            InstructionPB.Instruction inst = fragments.getAssembly();
            if (inst.getProtocolVersion() != NetworkConstants.MOSH_PROTOCOL_VERSION) {
                throw new NetworkException("mosh protocol version mismatch");
            }

            instQueue.offer(inst);
        }
    }

    @DiffThread
    public void pushBackEvent(UserEvent event) {
        sender.pushBackEvent(event);
    }

    @DiffThread
    public void tick() {
        recv();
        sender.tick();
    }

    public Queue<byte[]> getOutputQueue() {
        return ((CompleteTerminal) state).getOutputQueue();
    }

    private void recv() {
        while (!instQueue.isEmpty()) {
            InstructionPB.Instruction inst = instQueue.poll();

            sender.processAcknowledgmentThrough(inst.getAckNum());

            /* 1. make sure we don't already have the new state */
            for (TimestampedState<RemoteState> state : receiveStates) {
                if (inst.getNewNum() == state.num) {
                    return;
                }
            }
            /* 2. make sure we do have the old state */
            boolean found = false;
            for (TimestampedState<RemoteState> state : receiveStates) {
                if (inst.getOldNum() == state.num) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return;
            }

            processThrowawayUntil(inst.getThrowawayNum());

            TimestampedState<RemoteState> newState = new TimestampedState<>();
            newState.state = state;
            newState.timestamp = Timestamp.timestamp();
            newState.num = inst.getNewNum();

            byte[] diff = inst.getDiff().toByteArray();
            boolean dataAcked = false;
            if (diff != null && diff.length > 0) {
                info("Receive packet oldNum = {}, newNum = {}, ackNum = {}, throwawayNum = {}, diff = {}",
                        inst.getOldNum(), inst.getNewNum(), inst.getAckNum(), inst.getThrowawayNum(), inst.getDiff().toStringUtf8());
                if (inst.getAckNum() == 2) {
                    return;
                }
                newState.state.applyString(diff, inst.getAckNum());

                dataAcked = true;
            }

            for (int i = 0; i < receiveStates.size(); i++) {
                TimestampedState<RemoteState> state = receiveStates.get(i);
                if (state.num > newState.num) {
                    receiveStates.add(i, newState);
                    warn("Received OUT-OF-ORDER state {} [ack {}]", newState.num, inst.getAckNum());
                    return;
                }
            }

            receiveStates.add(newState);
            sender.setAckNum(newState.num);

            sender.remoteHeard(newState.timestamp);
            if (dataAcked) {
                sender.setDataAck();
            }
        }
    }

    private void processThrowawayUntil(long throwawayNum) {
        // when sender's throwaway num equals receiver's ackNum there were problems
        receiveStates.removeIf(next -> next.num < throwawayNum);
        assert receiveStates.size() > 0;
    }

    public record Addr(String serverHost, int port, String key) {
        public String serverHost() {
            return serverHost;
        }

        public int port() {
            return port;
        }

        public String key() {
            return key;
        }
    }

}

File: mosh/src/network/transport.rs
#![allow(dead_code)]

use super::{
    Connection, Fragment, FragmentAssembly, TimestampState, TransportSender, MOSH_PROTOCOL_VERSION,
};
use crate::statesync::{CompleteTerminal, UserStream, UserEvent};
use log::{error, warn};
use std::{cell::RefCell, rc::Rc};
use utilities::TimeStamp;

pub struct Transport {
    last_receive_state: CompleteTerminal,
    /* the underlying, encrypted network connection */
    connection: Rc<RefCell<Connection>>,
    /* sender side */
    sender: TransportSender,

    fragments: FragmentAssembly,
    receive_states: Vec<Rc<RefCell<TimestampState<CompleteTerminal>>>>,
}

impl Transport {
    pub fn new(
        initial_state: UserStream,
        initial_remote: CompleteTerminal,
        ip: &str,
        port: &str,
        key: &str,
    ) -> Self {
        let connection = Rc::new(RefCell::new(Connection::new(ip, port, key)));
        let mut receive_states = vec![];
        receive_states.push(Rc::new(RefCell::new(TimestampState::new(
            TimeStamp::timestamp(),
            0,
            initial_remote.clone(),
        ))));
        Transport {
            last_receive_state: initial_remote,
            connection: connection.clone(),
            sender: TransportSender::new(initial_state, connection),
            fragments: FragmentAssembly::new(),
            receive_states,
        }
    }

    pub fn tick(&mut self) {
        self.sender.tick();
    }

    pub fn push_back_event(&mut self, event: UserEvent) {
        self.sender.push_back_event(event);
    }

    pub fn recv(&self) -> Option<Vec<u8>> {
        self.connection.borrow().recv()
    }

    pub fn receive_packet(&mut self, bytes: Vec<u8>) {
        let bytes = self.connection.borrow_mut().recv_one(bytes);
        let fragment = Fragment::from_bytes(&bytes);
        if self.fragments.add_fragment(fragment) {
            if let Some(inst) = self.fragments.get_assembly() {
                if inst.protocol_version() != MOSH_PROTOCOL_VERSION {
                    error!(
                        "Mosh protocol version mismatch, accept = {}, get = {}",
                        MOSH_PROTOCOL_VERSION,
                        inst.protocol_version()
                    );
                    return;
                }

                self.sender.process_acknowledgment_through(inst.ack_num());

                /* 1. make sure we don't already have the new state */
                for state in self.receive_states.iter() {
                    if inst.new_num() == state.borrow().num {
                        return;
                    }
                }
                /* 2. make sure we do have the old state */
                let mut reference_state: Option<Rc<RefCell<TimestampState<CompleteTerminal>>>> =
                    None;
                let mut found = false;
                for state in self.receive_states.iter() {
                    if inst.old_num() == state.borrow().num {
                        reference_state = Some(state.clone());
                        found = true;
                        break;
                    }
                }
                if !found {
                    return;
                }

                self.process_throwaway_until(inst.throwaway_num());

                let mut new_state = TimestampState::new(
                    TimeStamp::timestamp(),
                    inst.new_num(),
                    reference_state.as_ref().unwrap().borrow().state.clone(),
                );

                let mut data_acked = false;
                if !inst.diff().is_empty() {
                    new_state.state.apply_string(inst.diff(), inst.ack_num());
                    data_acked = true;
                }

                for i in 0..self.receive_states.len() {
                    let state = self.receive_states.get(i);
                    if let Some(state) = state.as_deref() {
                        if state.borrow().num > new_state.num {
                            warn!(
                                "Received OUT-OF-ORDER state {} [ack {}]",
                                new_state.num,
                                inst.ack_num()
                            );
                            self.receive_states
                                .insert(i, Rc::new(RefCell::new(new_state)));
                            return;
                        }
                    }
                }

                self.sender.set_ack_num(new_state.num);
                self.sender.remote_heard(new_state.timestamp);
                self.receive_states.push(Rc::new(RefCell::new(new_state)));

                if data_acked {
                    self.sender.set_data_ack();
                }
            }
        }
    }

    fn process_throwaway_until(&mut self, throwaway_num: u64) {
        // when sender's throwaway num equals receiver's ackNum there were problems
        self.receive_states
            .retain(|state| state.borrow().num >= throwaway_num);
        assert!(self.receive_states.len() > 0);
    }
}
